File: src/karaoke/cdgRenderer.ts
// cdgRenderer.ts
// Renders CDG packets to a Canvas 2D context

export class CDGRenderer {
  width = 300;
  height = 216;

  // CDG uses a 16-color palette
  palette: string[] = new Array(16).fill("#000000");

  // Pixel buffer
  buffer: Uint8ClampedArray;

  ctx: CanvasRenderingContext2D;
  imageData: ImageData;

  constructor(canvas: HTMLCanvasElement) {
    this.ctx = canvas.getContext("2d")!;
    canvas.width = this.width;
    canvas.height = this.height;

    this.imageData = this.ctx.createImageData(this.width, this.height);
    this.buffer = this.imageData.data;
  }

  clear() {
    this.buffer.fill(0);
    this.ctx.putImageData(this.imageData, 0, 0);
  }

  setColorTable(data: Uint8Array) {
    for (let i = 0; i < 16; i++) {
      const r = (data[i * 2] & 0x3F) << 2;
      const g = ((data[i * 2] >> 6) | ((data[i * 2 + 1] & 0x0F) << 2)) << 2;
      const b = (data[i * 2 + 1] >> 4) << 4;

      this.palette[i] = `rgb(${r}, ${g}, ${b})`;
    }
  }

  drawTile(x: number, y: number, color: number) {
    const idx = (y * this.width + x) * 4;
    const [r, g, b] = this.palette[color]
      .replace("rgb(", "")
      .replace(")", "")
      .split(",")
      .map((v) => parseInt(v.trim()));

    this.buffer[idx] = r;
    this.buffer[idx + 1] = g;
    this.buffer[idx + 2] = b;
    this.buffer[idx + 3] = 255;
  }

  applyPacket(packet: any) {
    const { instruction, data } = packet;

    switch (instruction) {
      case 1: // Memory preset
        this.clear();
        break;

      case 6: // Tile block
      case 38: // Tile block XOR
        this.drawTileBlock(data);
        break;

      case 30: // Color table low
      case 31: // Color table high
        this.setColorTable(data);
        break;

      default:
        break;
    }

    this.ctx.putImageData(this.imageData, 0, 0);
  }

  drawTileBlock(data: Uint8Array) {
    const color = data[0] & 0x0F;
    const row = data[1] & 0x1F;
    const col = data[2] & 0x3F;

    const x = col * 6;
    const y = row * 12;

    let idx = 3;
    for (let r = 0; r < 12; r++) {
      const byte = data[idx++];
      for (let bit = 0; bit < 6; bit++) {
        if (byte & (1 << (5 - bit))) {
          this.drawTile(x + bit, y + r, color);
        }
      }
    }
  }
}

File: src-tauri/src/karaoke/cdg_renderer.rs
// src-tauri/src/karaoke/cdg_renderer.rs

use crate::karaoke::cdg_decoder::CDGPacket;


#[derive(Clone, Debug)]
pub struct CDGRenderer {
    pub width: usize,
    pub height: usize,
    pub buffer: Vec<u8>,
    pub palette: [[u8; 3]; 16],
}

impl CDGRenderer {
    pub fn new() -> Self {
        Self {
            width: 300,
            height: 216,
            buffer: vec![0; 300 * 216 * 3],
            palette: [[0, 0, 0]; 16],
        }
    }

    pub fn apply_packet(&mut self, packet: &CDGPacket) {
        match packet.instruction {
            1 => self.clear(),
            6 | 38 => self.draw_tile_block(&packet.data),
            30 | 31 => self.set_color_table(&packet.data),
            _ => {}
        }
    }

    fn clear(&mut self) {
        for px in self.buffer.iter_mut() {
            *px = 0;
        }
    }

    fn set_color_table(&mut self, data: &[u8; 16]) {
        for i in 0..16 {
            let r = (data[i] & 0x3F) << 2;
            let g = ((data[i] >> 6) | ((data[i] & 0x0F) << 2)) << 2;
            let b = (data[i] >> 4) << 4;

            self.palette[i] = [r, g, b];
        }
    }

    fn draw_tile_block(&mut self, data: &[u8; 16]) {
        let color = data[0] & 0x0F;
        let row = data[1] & 0x1F;
        let col = data[2] & 0x3F;

        let x = (col as usize) * 6;
        let y = (row as usize) * 12;

        let mut idx = 3;
        for r in 0..12 {
            let byte = data[idx];
            idx += 1;

            for bit in 0..6 {
                if byte & (1 << (5 - bit)) != 0 {
                    self.draw_pixel(x + bit, y + r, color);
                }
            }
        }
    }

    fn draw_pixel(&mut self, x: usize, y: usize, color: u8) {
        let offset = (y * self.width + x) * 3;
        let [r, g, b] = self.palette[color as usize];
        self.buffer[offset] = r;
        self.buffer[offset + 1] = g;
        self.buffer[offset + 2] = b;
    }
}
